core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```diff
@@ -0,0 +1,212 @@
+        .order(ByteOrder.LITTLE_ENDIAN).getInt();
+    cipherBlockSize = plainBlockSize + AesGcmOutputStream.GCM_NONCE_LENGTH + AesGcmOutputStream.GCM_TAG_LENGTH;
+
+    try {
+      gcmCipher = Cipher.getInstance("AES/GCM/NoPadding");
+    } catch (GeneralSecurityException e) {
+      throw new IOException(e);
+    }
+    this.nonce = new byte[AesGcmOutputStream.GCM_NONCE_LENGTH];
+    this.key = new SecretKeySpec(aesKey, "AES");
+    this.ciphertextBlockBuffer = new byte[cipherBlockSize];
+    this.currentBlockIndex = 0;
+    this.currentOffsetInPlainBlock = 0;
+
+    numberOfBlocks = (int) (netSourceFileSize / cipherBlockSize);
+    lastBlockSize = (int) (netSourceFileSize % cipherBlockSize);
+    if (lastBlockSize == 0) {
+      lastBlockSize = cipherBlockSize;
+    } else {
+      numberOfBlocks += 1;
+    }
+
+    plainStreamSize = (numberOfBlocks - 1L) * plainBlockSize +
+            (lastBlockSize - AesGcmOutputStream.GCM_NONCE_LENGTH - AesGcmOutputStream.GCM_TAG_LENGTH);
+  }
+
+  public long plaintextStreamSize() {
+    return plainStreamSize;
+  }
+
+  @Override
+  public int available() throws IOException {
+    return Math.toIntExact(plainStreamSize - plainStreamPosition);
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    if (len <= 0) {
+      throw new IOException("Negative read length " + len);
+    }
+
+    if (available() <= 0) {
+      return -1;
+    }
+
+    boolean lastBlock = currentBlockIndex + 1 == numberOfBlocks;
+    int resultBufferOffset = off;
+    int remaining = len;
+
+    sourceStream.seek(AesGcmOutputStream.PREFIX_LENGTH + currentBlockIndex * cipherBlockSize);
+
+    while (remaining > 0) {
+      int toLoad = lastBlock ? lastBlockSize : cipherBlockSize;
+      int loaded = sourceStream.read(ciphertextBlockBuffer, 0, toLoad);
+      if (loaded != toLoad) {
+        throw new IOException("Read " + loaded + " instead of " + toLoad);
+      }
+
+      // Copy nonce
+      System.arraycopy(ciphertextBlockBuffer, 0, nonce, 0, AesGcmOutputStream.GCM_NONCE_LENGTH);
+
+      byte[] aad = AesGcmOutputStream.calculateAAD(fileAadPrefix, currentBlockIndex);
+      byte[] plaintextBlock;
+      try {
+        GCMParameterSpec spec = new GCMParameterSpec(AesGcmOutputStream.GCM_TAG_LENGTH_BITS, nonce);
+        gcmCipher.init(Cipher.DECRYPT_MODE, key, spec);
+        gcmCipher.updateAAD(aad);
+
+        plaintextBlock = gcmCipher.doFinal(ciphertextBlockBuffer, AesGcmOutputStream.GCM_NONCE_LENGTH,
+                toLoad - AesGcmOutputStream.GCM_NONCE_LENGTH);
+      } catch (GeneralSecurityException e) {
+        throw new IOException("Failed to decrypt", e);
+      }
+
+      int remainingInBlock = plaintextBlock.length - currentOffsetInPlainBlock;
+      boolean finishTheBlock = remaining >= remainingInBlock;
```

Comment:
Not sure I understand this? remainingInBlock is unencrypted bytes and remaining is encrypted right?

Reply:
remaining is also unencrypted. If it's longer than what's left in current block, we'll read the block to the end